include/deal.II/hp/fe_collection.h

```diff
@@ -1006,13 +1006,11 @@ namespace hp
   FECollection<dim, spacedim>::hp_constraints_are_implemented() const
   {
     Assert(finite_elements.size() > 0, ExcNoFiniteElements());
-
-    bool hp_constraints = true;
-    for (unsigned int i = 0; i < finite_elements.size(); ++i)
-      hp_constraints =
-        hp_constraints && finite_elements[i]->hp_constraints_are_implemented();
-
-    return hp_constraints;
+    return std::all_of(finite_elements.cbegin(),
+                       finite_elements.cend(),
+                       [](decltype(*finite_elements.cbegin()) fe) {
```

Comment:
we need C++14 template lambdas! :-)

Reply:
But you can make it more readable this way, because I'm pretty sure that the iterators should be convertible:
```suggestion
                       [](const std::shared_ptr<FiniteElement<dim,spacedim>> &fe) {
```

Reply:
> we need C++14 template lambdas! :-)

You mean we should move to C++14 soonish (like after the release)? :wink:

Reply:
`const auto &fe` would've been great for here, but with `decltype` I tried to write it in the most generic way. My thinking was that if we change the exact type signature of `finite_elements` (for example from `vector<shared_ptr>` to `vector<unique_ptr>`) the lambda should work properly.

I understand that using the full type name in the argument list is probably the most readable form. So I will apply @bangerth's suggestion.

Reply:
> You mean we should move to C++14 soonish (like after the release)?

Well well, now, let's not get crazy! ;-))

Reply:
Thanks -- we've generally tried to be explicit about types where possible. It's just one way to make it easier for readers to reason about the correctness of code. (Within limits where the type is obvious -- e.g., in `for (const auto &cell : ...)`.)